File: mobile-ios/treasuryflow/Data/Repository/ExampleRepository.swift
import Foundation

protocol ExampleRepository: Sendable {
    func list(page: Int, size: Int) async throws -> ExampleListResponse
    func getById(_ id: String) async throws -> ExampleResponse
    func create(title: String, description: String?) async throws -> ExampleResponse
    func update(id: String, title: String?, description: String?, status: String?) async throws -> ExampleResponse
    func delete(_ id: String) async throws
}

final class ExampleRepositoryImpl: ExampleRepository, Sendable {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func list(page: Int = 0, size: Int = 20) async throws -> ExampleListResponse {
        try await apiClient.request(.listExamples(page: page, size: size))
    }

    func getById(_ id: String) async throws -> ExampleResponse {
        try await apiClient.request(.getExample(id: id))
    }

    func create(title: String, description: String?) async throws -> ExampleResponse {
        let body = CreateExampleRequest(title: title, description: description)
        return try await apiClient.request(.createExample(), body: body)
    }

    func update(id: String, title: String?, description: String?, status: String?) async throws -> ExampleResponse {
        let body = UpdateExampleRequest(title: title, description: description, status: status)
        return try await apiClient.request(.updateExample(id: id), body: body)
    }

    func delete(_ id: String) async throws {
        try await apiClient.requestVoid(.deleteExample(id: id))
    }
}

File: mobile-android/app/src/main/kotlin/com/mortitech/treasuryflow/feature/example/data/repository/ExampleRepository.kt
package com.mortitech.treasuryflow.feature.example.data.repository

import com.mortitech.treasuryflow.core.common.AppResult
import com.mortitech.treasuryflow.core.common.appResultOf
import com.mortitech.treasuryflow.core.model.PagedData
import com.mortitech.treasuryflow.core.network.ApiService
import com.mortitech.treasuryflow.feature.example.data.remote.dto.CreateExampleRequest
import com.mortitech.treasuryflow.feature.example.data.remote.dto.UpdateExampleRequest
import com.mortitech.treasuryflow.feature.example.domain.model.Example
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ExampleRepository @Inject constructor(
    private val apiService: ApiService,
) {

    suspend fun list(page: Int = 0, size: Int = 20): AppResult<PagedData<Example>> =
        appResultOf {
            val response = apiService.listExamples(page = page, size = size)
            PagedData(
                content = response.content.map { it.toDomain() },
                page = response.page,
                size = response.size,
                totalElements = response.totalElements,
                totalPages = response.totalPages,
            )
        }

    suspend fun getById(id: String): AppResult<Example> =
        appResultOf {
            apiService.getExample(id).toDomain()
        }

    suspend fun create(title: String, description: String?): AppResult<Example> =
        appResultOf {
            apiService.createExample(CreateExampleRequest(title, description)).toDomain()
        }

    suspend fun update(id: String, title: String?, description: String?, status: String?): AppResult<Example> =
        appResultOf {
            apiService.updateExample(id, UpdateExampleRequest(title, description, status)).toDomain()
        }

    suspend fun delete(id: String): AppResult<Unit> =
        appResultOf {
            apiService.deleteExample(id)
        }
}
